File: message-processor-reverse/src/main/java/message/processor/reverse/processor/MessageProcessorExecutorService.java
package message.processor.reverse.processor;

import message.core.wrapper.MessageWrapperQueue;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import javax.annotation.PreDestroy;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@Service
public class MessageProcessorExecutorService {

    private final ExecutorService executorService;

    @Autowired
    public MessageProcessorExecutorService(@Qualifier("processor-reverse") MessageWrapperQueue messageProcessorQueue,
                                           MessageWrapperQueue messageDispatcherQueue,
                                           @Value("${consumers.size:1}") Integer consumerSize) {

        ExecutorService executorService = Executors.newFixedThreadPool(consumerSize);
        for (int i = 0; i < consumerSize; i++) {
            MessageProcessorQueueConsumer processor = new MessageProcessorQueueConsumer(messageProcessorQueue, messageDispatcherQueue);
            executorService.submit(processor);
        }
        this.executorService = executorService;
    }

    @PreDestroy
    public void deinit() {
        executorService.shutdown();
    }

}

File: message-processor-echo/src/main/kotlin/message/processor/echo/processor/MessageProcessorExecutorService.kt
package message.processor.echo.processor

import message.core.log.Log
import message.core.wrapper.MessageWrapperQueue
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import javax.annotation.PreDestroy

@Service
class MessageProcessorExecutorService @Autowired
constructor(@Qualifier("processor-echo") messageEchoProcessorQueue: MessageWrapperQueue,
            messageDispatcherQueue: MessageWrapperQueue,
            @Value("\${consumers.size:1}") consumerSize: Int?) {

    private val executorService: ExecutorService

    init {
        Log.application.info("Creating {} consumers to process echo", consumerSize)
        val executorService = Executors.newFixedThreadPool(consumerSize!!)
        for (i in 0 until consumerSize) {
            val processor = MessageProcessorQueueConsumer(messageEchoProcessorQueue, messageDispatcherQueue)
            executorService.submit(processor)
        }
        this.executorService = executorService
    }

    @PreDestroy
    fun deinit() {
        executorService.shutdown()
    }

}
